lib/block.mli

```diff
@@ -30,26 +40,68 @@ type section = int * string
 (** The type for sections. *)
 
 (** The type for supported code blocks. *)
-type t = {
+type ocaml_block = {
+  line    : int;
+  file    : string;
+  section : section option;
+  labels  : Label.t list;
+  header  : Header.t option;
+  contents: string list;
+  value   : value;
+  non_det : [`Output | `Command] option;
+  env     : string;
+  (** [env] is the name given to the environment where tests are run. *)
+}
+
+type shell_block = {
   line    : int;
   file    : string;
   section : section option;
   labels  : Label.t list;
-  header  : string option;
+  header  : Header.t option;
   contents: string list;
   value   : value;
+  non_det : [`Output | `Command] option;
 }
 
-val empty: t
-(** [empty] is the empty block. *)
+type include_block = {
+  line : int;
+  file : string;
+  section : section option;
+  labels : Label.t list;
+  header : Header.t option;
+  contents : string list;
+  value : value;
+  file_included : string;
+  (** [file_included] is the name of the file to synchronize with. *)
+  part_included : string option;
+  (** [part_included] is the part of the file to synchronize with.
+      If lines is not specified synchronize the whole file. *)
+}
+
+type t =
+  | Toplevel of ocaml_block
+  | Shell of shell_block
+  | Include of include_block
+  | OCaml of ocaml_block
```

Comment:
What is the advantage of branching here rather than with the `value` field, as before (it's still here btw) ?
What does the `value` field mean now and can it be set to inconsistent values ?

I think we should not work around the `value` type and the way it works, but instead improve it. 
For example:
```diff
-type cram_value = { pad: int; tests: Cram.t list }
+type cram_value = {
+  pad: int;
+  tests: Cram.t list;
+  non_det : [`Output | `Command] option;
+}
+
+type ocaml_value = {
+  non_det : [`Output | `Command] option;
+  env     : string;
+  (** [env] is the name given to the environment where tests are run. *)
+}
+
+type include_value = {
+  file_included : string;
+  (** [file_included] is the name of the file to synchronize with. *)
+  part_included : string option;
+  (** [part_included] is the part of the file to synchronize with.
+      If lines is not specified synchronize the whole file. *)
+}
 
 (** The type for block values. *)
 type value =
   | Raw
-  | OCaml
+  | OCaml of ocaml_value
   | Error of string list
   | Cram of cram_value
   | Toplevel of Toplevel.t list
+  | Include of include_value
```

Reply:
A partial reason for this is that the `value` is only computed when the block is evaluated, and not when the block is built (the value is only set to `Raw`), I will have a look if that's possible to move this evaluation during the parsing.

Reply:
I think we shouldn't duplicate every single field of the main type but instead have the shared part in a record and the block type specific parts as record arguments of a variant type, as @Julow  suggest although value probably isn't the right name here.

The main reason for this change is that in the final type we can get rid of the unproper `Label.t list` and have fields for each valid label instead.

Reply:
I do believe we should move `eval` at parsing time indeed!

Reply:
If `eval` has to be a separate step, we should let the previous type `t` as it was, as the "parsed blocks". `eval` should return a richer type instead.
Example:
- `value` should only be informations available a parsing, it should not allows dummy values until `eval` is called.
  `t` may not even need a `value` field
- `eval` should return something like the new `type t` you made, without the "parsed block" bits
  Example:
  ```
  type evaluated = OCaml of { env : Env.t } | Toplevel of Toplevel.t list | Include of { file_included : ... } | ...
  ```
  We can pass `t`s and `evaluated`s around when we need without constraints on "did it pass through `eval`" or "is it a valid block"

Reply:
Yeah to clarify I think we should have two different types, `raw` which can hold inconsistent values and should be fairly similar to current type `t`, and `t` which is the clean type ie a type that doesn't allow inconsistent value.

Rough parsing returns `raw` values which are then turned into `t`. That step is currently called `eval` although it doesn't evaluate anything and should probably be renamed.

Everywhere else in the code, we should use `t` and never `raw`. `raw` is just an intermediate representation.

Reply:
I improved this, the evaluation is now done during parsing, but I didn't find a better name for the `value` field.